service/docs/source/geopm_agent_frequency_balancer.7.rst

```diff
@@ -0,0 +1,132 @@
+
+The following subsections outline the agent's responsibilities at a high level.
+See (**TODO** get a paper reference. Also add to the publications page.) for more details.
+
+Critical Path Time Estimates
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The agent measures the non-networking time as time that each core spends with
+the ``GEOPM_REGION_HINT_NETWORK`` hint *inactive* in each epoch. By default,
+GEOPM annotates all MPI function calls with that hint. The agent aims to set
+frequency controls that reduce the variation of non-networking time across
+cores.
+
+Frequency Control Decisions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Frequency control decisions applied to one core may impact the achievable
+frequency on other cores. In general, if the platform throttles frequency
+lower than a user's requested frequency due to resource constraints, then
+throttling other cores may lessen the need for platform-induced throttling.
+Furthermore some platforms support heterogeneous turbo frequency limits, where
+the number of cores configured with high turbo limits impacts the maximum
+achievable frequency. The agent performs estimates of the application
+performance tradeoffs in these cases.
+
+Heterogeneous Turbo Frequency Limits
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Turbo frequency boosting allows a CPU to temporarily achieve higher than its
+base frequency while still respecting hardware design constraints. Typically,
+higher turbo frequency is achievable when only one CPU core is active (i.e.,
+*single-core turbo*) than when all CPU cores are active (i.e., *all-core
+turbo*). Some platforms support heterogeneous user-configurable turbo frequency
+limits through the *Intel(R) Speed Select Technology - Turbo Frequency*
+(SST-TF) feature.
+
+This agent queries the platform to learn the frequency trade-offs of different
+possible SST-TF configurations. The agent estimates the potential critical
+path time for each possible configuration, and applies the configuration that
+results in the least estimated time.
+
+Agent Configuration
+-------------------
+Execute an application with this agent by specifying
+``--geopm-agent=frequency_balancer`` in the :doc:`geopmlaunch(1) <geopmlaunch.1>`
+command line arguments.
+
+The agent is capable of influencing CPU core performance through P-State
+settings and through heterogeneous turbo frequency limits via SST-TF. Each
+type of control can be used in isolation, or they can be used together.
+If SST-TF is requested on a platform that does not support that feature, then
+the request to use SST-TF is ignored.
+
+Specify the desired types of performance influencers through a policy json file
+and indicate the location of the file with ``--geopm-policy=<path-to-file>``.
+The :doc:`geopmagent(1) <geopmagent.1>` tool can help construct a policy file.
+
+Agent Policy Definitions
+------------------------
+This agent's behavior is influenced by the following GEOPM policy fields. At
+least one balancing option must be set to ``1``.
```

Comment:
What happens if both policy fields are set to 1?  Add a line of text for this please.